Issue 64

silver-1347 opened the pull request:
`participant` の `schedules` 及び `events` が空の場合のみボタンを表示するようにした
そうでない場合は退会できない旨を代わりに表示

silver-1347 committed:
[ref #64] 予定が入っている被験者は退会できないようにする

participantがschedulesもeventsも持ってない時にアカウント削除ボタンを表示するように修正
そうでない場合は削除できない旨を表示

silver-1347 committed:
[ref #64] 予定が入っている被験者は退会できないようにする

条件分岐のミスを修正

silver-1347 committed:
[ref #64] 予定が入っている被験者は退会できないようにする

退会できる場合とできない場合でテストに条件分岐を設定

okabi commented on app/views/participants/_destroy.html.erb:
```diff
@@ -1,7 +1,11 @@
-                    commit:  '削除する',
-                    cancel:  'キャンセル', },
-            class: 'form-control btn btn-danger' %>
+<% if participant.schedules.blank? && participant.events.blank? %>
```
`participant.schedules.blank?` だと、実施済み・拒否されたスケジュールも判定してしまうので、
現在日時と `application.status` の状態も考慮したいです。
`event` についても現在日時から判定したいです。

okabi commented on app/views/participants/_destroy.html.erb:
```diff
@@ -1,7 +1,11 @@
+                        cancel:  'キャンセル', },
+                        class: 'form-control btn btn-danger' %>
+<% else%>
+    <p> あなたはまだ終了していない応募済みの実験があるためアカウントを削除できません。</p>
```
ちょっと表現が冗長な気がするので、
「予定が残っているためアカウントを削除できません」くらいにしたい気がします。
あと、 `disabled=true` なボタンを表示して、将来的にそのボタンから削除できることを示しておきたいです。

okabi commented on test/integration/participants_deactivate_test.rb:
```diff
@@ -17,7 +17,11 @@ def deactivate
     log_in_as_participant @participant
     get settings_path
-    assert_select "a[href=?]", deactivate_path
+    if @participant.schedules.blank? && @participant.events.blank?
```
app/views/participants/_destroy.html.erb のコメントと同様

okabi reviewed (changes requested):
`deactivate_path` に飛んだ時のコントローラでも、
アカウントを無効化して良い被験者か判定したいです。

silver-1347 committed:
[ref #64] 予定が入っている被験者は退会できないようにする

participantにdeactivateできるかを判別するメソッドを追加
削除できないアカウントの場合にdisabledなボタンを表示するように修正
controllerでアカウント削除して良いかを判定

okabi commented on app/models/participant.rb:
```diff
@@ -70,4 +70,22 @@ def self.gender_to_s(gender)
     Participant.find_by(email: email, deactivated: false)
   end
+
+  def deactivatable?
```
良いモジュール化だと思います 👍

okabi commented on test/integration/participants_deactivate_test.rb:
```diff
@@ -17,15 +17,19 @@ def deactivate
-    assert @participant.deactivated
-    get login_path
-    assert flash.empty?
+    if @participant.deactivatable?
```
これだと `@participant` が `deactivatable` なパターンかそうじゃないパターンのどちらかしかテストできないです。。。
両方ともテストしておきたいです。

silver-1347 replied:
ここの修正中に気づいたのですが、 `Schedule` のテストデータが現在から見ると全て終了していることになるので2パターン両方のテストができない状態です。何かいい方法ありませんか？

yegawa replied:
ymlファイルをyml.erbにして、今日の日付で1時間前から23時59分まで実験するやつとか（別に24でも良いけど日付またぐ処理書くのだるそう＆実際使わんしなあ）
将来の日付で実験するやつとか書くといいかも？

http://qiita.com/itkrt2y/items/ca34fea17fc7dde56b7a

okabi replied:
2100年の予定とかぶっ込んでみると良いかも。。。正しく扱えるのか知らないけど。

okabi reviewed (changes requested):
テストの修正だけお願いします。

silver-1347 committed:
[ref #64] 予定が入っている被験者は退会できないようにする

退会できるアカウントと退会できないアカウントの2パターンのテストを用意
しばらくは終了しないスケジュールやイベントのテストデータを追加
